aim/agent/aid/universes/k8s/k8s_watcher.py

```diff
@@ -179,40 +179,75 @@ def _observe_and_monitor_loop(self):
             raise exc
         time.sleep(MONITOR_LOOP_MAX_WAIT)
 
+    def _init_aim_k8s(self, types_to_observe):
+        if self._needs_init:
+            # NOTE(ivar): we need to lock the observer here to prevent it
+            # from reading empty or incomplete trees
+            # REVISIT(ivar): this is NOT gonna work for multi AID. In general,
+            # the whole K8S watcher cannot run as-is in a multi AID environment
+            with utils.get_rlock(lcon.AID_OBSERVER_LOCK):
+                self._reset_trees()
+                self._renew_klient_watch()
+
+                self._version_by_type = {}
+                for typ in self._k8s_types_to_observe:
+                    self._init_stream_for_type(typ)
+                self._persistence_loop(save_on_empty=True)
+                LOG.info("Trees initialized")
+                self._needs_init = False
+
+    @utils.rlock(lcon.K8S_WATCHER_TREE_LOCK)
     def _start_observers(self, types_to_observe):
-        self._reset_trees()
-        self._renew_klient_watch()
-
-        self._observe_thread_state = {}
+        self._init_aim_k8s(types_to_observe)
         for id, typ in enumerate(list(types_to_observe)):
             self._observe_thread_state[id] = dict(watch_stop=False)
-            thd = eventlet.spawn(self._observe_objects, typ, id)
+            thd = utils.spawn_thread(
+                self._observe_objects, self.klient.watch.stream, typ, id,
+                self._version_by_type.get(typ))
             self._observe_thread_state[id]['thread'] = thd
 
+    def _init_stream_for_type(self, k8s_type):
+
+        def wrapped_list(f, k8s_type, *args, **kwargs):
+            result = []
+            list = self.klient.list(k8s_type, namespace=kwargs['namespace'])
+            self._version_by_type[k8s_type] = list[
+                'metadata']['resourceVersion']
+            for item in list['items']:
+                result.append({'type': ACTION_CREATED,
+                               'raw_object': item,
+                               'object': item})
+            return result
+        self._observe_objects(wrapped_list, k8s_type, None, None)
+
     def _check_observers(self):
         exc = None
         for t in self._observe_thread_state:
             tstate = self._observe_thread_state[t]
-            thd = tstate['thread']
-            if tstate.get('watch_exception'):
-                exc = tstate.get('watch_exception')
-                LOG.info('Thread %s raised exception %s', thd, exc)
-                break
-            if tstate.get('http_resp') and tstate['http_resp'].closed:
-                LOG.info('HTTP response closed for thread %s', thd)
-                exc = K8SObserverStopped()
-                break
-            if thd.dead:
-                LOG.info('Thread %s is not alive', thd)
-                exc = K8SObserverStopped()
-                break
+            thd = tstate.get('thread')
+            LOG.info("Checking thread %s for %s" % (thd, t))
```

Comment:
oh this was something I did for debug and somehow ended up in this branch, will clean up